label_sync/main.go

```diff
@@ -0,0 +1,439 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This is a label_sync tool, details in README.md
+package main
+
+import (
+	"bytes"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/ghodss/yaml"
+
+	"k8s.io/test-infra/prow/github"
+)
+
+// Label single label
+type Label struct {
+	Name  string `json:"name"`
+	Color string `json:"color"`
+}
+
+// RequiredLabels is a list of Required Labels to sync in all kubernetes repos
+type RequiredLabels struct {
+	Labels []Label `json:"labels"`
+}
+
+// RepoList is a list of Repositories from Org to be checked by this program
+type RepoList struct {
+	Org   string        `json:"org"`
+	Repos []github.Repo `json:"repos"`
+}
+
+// RepoLabels is a mapping repository name --> current list of labels
+type RepoLabels struct {
+	Labels map[string][]github.Label
+}
+
+// UpdateItem is a single item to update
+// update can mean: update Label in repo or add missing label in repo
+type UpdateItem struct {
+	Why                         string
+	RequiredLabel, CurrentLabel Label
+}
+
+// UpdateData Repositories to update: map repo name --> list of Updates
+type UpdateData struct {
+	ReposToUpdate map[string][]UpdateItem
+}
+
+var (
+	labelsPath         = flag.String("labels-path", "/etc/config/labels.yaml", "Path to labels.yaml.")
+	githubTokenFile    = flag.String("github-token-file", "/etc/github/oauth", "Path to the file containing the GitHub OAuth secret.")
+	org                = flag.String("org", "kubernetes", "Organization to fetch repository list from, to fetch user's public repositories, use user:user_name")
+	dry                = flag.Bool("dry-run", true, "Dry run for testing. Uses API tokens but does not mutate.")
+	local              = flag.Bool("local", false, "Run locally for testing purposes only. Does not require secret files.")
+	reposData          = flag.String("repos", "", "Repository file saved as YAML to save GitHub API points")
+	repoLabelsData     = flag.String("repo-labels", "", "Repository Labels file saved as YAML to save GitHub API points")
+	dumpReposData      = flag.String("dump-repos", "", "Save repositories found as YAML (so this file can be used again with -repos without using Gihub API points in -local mode)")
+	dumpRepoLabelsData = flag.String("dump-repo-labels", "", "Save repositories labels found as YAML (so this file can be used again with -repo-labels without using Gihub API points in -local mode)")
+	debug              = flag.Bool("debug", false, "Turn on debug to be more verbose")
```

Comment:
Generally speaking we've used [Cobra](https://github.com/spf13/cobra) in Kubernetes for command-line applications. I don't necessarily think it's better than the native flag system but it might be worth being consistent. If one of the other reviewers wants to chime in I'd appreciate it.

Reply:
Left untouched.